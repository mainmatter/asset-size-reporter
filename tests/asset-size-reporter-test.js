'use strict';

const path = require('path');
const stripAnsi = require('strip-ansi');

const report = require('../src/asset-size-reporter');

const FIXTURE_PATH = `${__dirname}/fixtures`;

class FakeConsole {
  constructor() {
    this.messages = [];
  }

  log(message = '') {
    this.messages.push(message);
  }

  get output() {
    return this.messages.join('\n');
  }
}

describe('asset-size-reporter', () => {
  describe('text reporter', () => {
    test('reports results as text by default', async () => {
      let cwd = path.join(FIXTURE_PATH, 'default');

      let patterns = [
        'dist/**/*.js',
        '!dist/ignored-*.js',
      ];

      let fakeConsole = new FakeConsole();

      await report({ patterns, cwd, console: fakeConsole });

      expect(stripAnsi(fakeConsole.output)).toMatchSnapshot();
    });

    test('supports `gzip: false`', async () => {
      let cwd = path.join(FIXTURE_PATH, 'default');

      let patterns = [
        'dist/**/*.js',
        '!dist/ignored-*.js',
      ];

      let fakeConsole = new FakeConsole();

      await report({ patterns, cwd, gzip: false, console: fakeConsole });

      expect(stripAnsi(fakeConsole.output)).toMatchSnapshot();
    });
  });

  describe('JSON reporter', () => {
    test('reports results as JSON if `json` option is used', async () => {
      let cwd = path.join(FIXTURE_PATH, 'default');

      let patterns = [
        'dist/**/*.js',
        '!dist/ignored-*.js',
      ];

      let fakeConsole = new FakeConsole();

      await report({ patterns, json: true, cwd, console: fakeConsole });

      let path1 = path.join('dist', 'file-inside-dist.js');
      let path2 = path.join('dist', 'foo', 'nested-file-inside-dist.js');

      expect(JSON.parse(fakeConsole.output)).toEqual({
        [path1]: { raw: 1855, gzip: 377, brotli: null },
        [path2]: { raw: 3075, gzip: 636, brotli: null },
      });
    });

    test('supports `gzip: false`', async () => {
      let cwd = path.join(FIXTURE_PATH, 'default');

      let patterns = [
        'dist/**/*.js',
        '!dist/ignored-*.js',
      ];

      let fakeConsole = new FakeConsole();

      await report({ patterns, json: true, gzip: false, cwd, console: fakeConsole });

      let path1 = path.join('dist', 'file-inside-dist.js');
      let path2 = path.join('dist', 'foo', 'nested-file-inside-dist.js');

      expect(JSON.parse(fakeConsole.output)).toEqual({
        [path1]: { raw: 1855, gzip: null, brotli: null },
        [path2]: { raw: 3075, gzip: null, brotli: null },
      });
    });
  });

  describe('compare reporter', () => {
    test('reports results as text by default', async () => {
      let cwd = path.join(FIXTURE_PATH, 'default');

      let patterns = [
        'dist/**/*.{js,txt}',
        '!dist/ignored-*.js',
      ];

      let path1 = path.join('dist', 'file-inside-dist.js');
      let path2 = path.join('dist', 'foo', 'nested-file-inside-dist.js');
      let path4 = path.join('dist', 'deleted-file-inside-dist.js');

      let compare = {
        [path1]: { raw: 1275, gzip: 144, brotli: null },
        [path2]: { raw: 3075, gzip: 636, brotli: null },
        [path4]: { raw: 4242, gzip: 42, brotli: null },
      };

      let fakeConsole = new FakeConsole();

      await report({ patterns, compare, cwd, console: fakeConsole });

      expect(stripAnsi(fakeConsole.output)).toMatchSnapshot();
    });

    test('supports `gzip: false`', async () => {
      let cwd = path.join(FIXTURE_PATH, 'default');

      let patterns = [
        'dist/**/*.js',
        '!dist/ignored-*.js',
      ];

      let path1 = path.join('dist', 'file-inside-dist.js');
      let path2 = path.join('dist', 'foo', 'nested-file-inside-dist.js');

      let compare = {
        [path1]: { raw: 1275, gzip: null, brotli: null },
        [path2]: { raw: 32075, gzip: null, brotli: null },
      };

      let fakeConsole = new FakeConsole();

      await report({ patterns, compare, gzip: false, cwd, console: fakeConsole });

      expect(stripAnsi(fakeConsole.output)).toMatchSnapshot();
    });
  });
});
