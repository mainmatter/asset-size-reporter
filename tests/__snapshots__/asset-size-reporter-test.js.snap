// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asset-size-reporter compare reporter reports results as text by default 1`] = `
"[dist/deleted-file-inside-dist.js]: 4.24 kB / 42 B gzip (deleted file)
dist/file-inside-dist.js: 1.27 kB -> 1.85 kB (+580 B) / 144 B -> 377 B (+233 B) gzip
dist/foo/nested-file-inside-dist.js: 3.08 kB / 636 B gzip
dist/text-file-in-dist.txt: 3.94 kB / 636 B gzip (new file)

Total: 8.59 kB -> 8.87 kB (+275 B) / 822 B -> 1.65 kB (+827 B) gzip"
`;

exports[`asset-size-reporter compare reporter supports \`gzip: false\` 1`] = `
"dist/file-inside-dist.js: 1.27 kB -> 1.85 kB (+580 B)
dist/foo/nested-file-inside-dist.js: 32.1 kB -> 3.08 kB (-29 kB)

Total: 33.4 kB -> 4.93 kB (-28.4 kB)"
`;

exports[`asset-size-reporter text reporter reports results as text by default 1`] = `
"dist/file-inside-dist.js: 1.85 kB / 377 B gzip
dist/foo/nested-file-inside-dist.js: 3.08 kB / 636 B gzip

Total: 4.93 kB / 1.01 kB gzip"
`;

exports[`asset-size-reporter text reporter supports \`gzip: false\` 1`] = `
"dist/file-inside-dist.js: 1.85 kB
dist/foo/nested-file-inside-dist.js: 3.08 kB

Total: 4.93 kB"
`;
